tonic/src/codec/mod.rs

```diff
@@ -37,3 +41,29 @@ pub trait Codec: Default {
     /// Fetch the decoder.
     fn decoder(&mut self) -> Self::Decoder;
 }
+
+/// Decoding of frames via buffers.
+pub trait Decoder {
+    /// The type of decoded frames.
+    type Item;
+
+    /// The type of unrecoverable frame decoding errors.
+    type Error: From<io::Error>;
+
+    /// Attempts to decode a frame from the provided buffer of bytes.
+    fn decode(&mut self, src: &mut DecodeBuf<'_>) -> Result<Option<Self::Item>, Self::Error>;
+}
+
+/// Trait of helper objects to write out messages as bytes.
+pub trait Encoder {
+    /// The type of items consumed by the `Encoder`
+    type Item;
+
+    /// The type of encoding errors.
+    ///
+    /// The type of unrecoverable frame encoding errors.
+    type Error: From<io::Error>;
+
+    /// Encodes a frame into the buffer provided.
+    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error>;
```

Comment:
I think for consitency we should do the same here for encode https://docs.rs/tower-grpc/0.1.1/src/tower_grpc/generic/codec.rs.html#135-137 and add an EncodeBuf which wraps bytesmut